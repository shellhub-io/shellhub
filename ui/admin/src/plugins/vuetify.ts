// Styles
// import "@mdi/font/css/materialdesignicons.css";
import "vuetify/styles";

// Vuetify
import { createVuetify } from "vuetify";

export default createVuetify({
  theme: {
    defaultTheme: "dark",
  },
});
