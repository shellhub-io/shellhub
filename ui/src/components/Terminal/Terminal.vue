<template>
  <div class="ma-0 pa-0 w-100 fill-height position-relative">
    <div ref="terminal" class="terminal" data-test="terminal-container" />
  </div>
</template>

<script setup lang="ts">
import { onMounted, onUnmounted, ref } from "vue";
import { useEventListener } from "@vueuse/core";
import "xterm/css/xterm.css";
import { Terminal } from "xterm";
import { FitAddon } from "xterm-addon-fit";
import { IParams } from "@/interfaces/IParams";
import { InputMessage, MessageKind, ResizeMessage, WebTermDimensions } from "@/interfaces/ITerminal";

const { token } = defineProps<{
  token: string;
}>();

const terminal = ref<HTMLElement>({} as HTMLElement);
const xterm = ref<Terminal>({} as Terminal);
const fitAddon = ref<FitAddon>(new FitAddon());
const ws = ref<WebSocket>({} as WebSocket);
const textEncoder = new TextEncoder();

const initializeTerminal = () => {
  xterm.value = new Terminal({
    cursorBlink: true,
    fontFamily: "monospace",
    theme: {
      background: "#0f1526",
    },
  });

  xterm.value.loadAddon(fitAddon.value);
};

const getWebTermDimensions = (): WebTermDimensions => ({
  cols: xterm.value.cols,
  rows: xterm.value.rows,
});

const encodeURLParams = (params: IParams): string => Object.entries(params).map(([key, value]) => `${key}=${value}`).join("&");

const isWebSocketOpen = () => ws.value.readyState === WebSocket.OPEN;

const getWebSocketUrl = (dimensions: WebTermDimensions): string => {
  const protocol = window.location.protocol === "http:" ? "ws" : "wss";
  const wsInfo = { token, ...dimensions };

  return `${protocol}://${window.location.host}/ws/ssh?${encodeURLParams(wsInfo)}`;
};

const setupTerminalEvents = () => {
  xterm.value.onData((data) => {
    if (!isWebSocketOpen()) return;

    const message: InputMessage = {
      kind: MessageKind.Input,
      data: [...textEncoder.encode(data)],
    };
    ws.value.send(JSON.stringify(message));
  });

  xterm.value.onResize((data) => {
    if (!isWebSocketOpen()) return;

    const message: ResizeMessage = {
      kind: MessageKind.Resize,
      data: { cols: data.cols, rows: data.rows },
    };
    ws.value.send(JSON.stringify(message));
  });
};

const setupWebSocketEvents = () => {
  ws.value.onopen = () => {
    fitAddon.value.fit();
  };

  ws.value.onmessage = (event) => {
    xterm.value.write(event.data);
  };

  ws.value.onclose = () => {
    xterm.value.write("\r\nConnection ended");
  };
};

const initializeWebSocket = () => {
  const dimensions = getWebTermDimensions();
  const wsUrl = getWebSocketUrl(dimensions);
  ws.value = new WebSocket(wsUrl);
  setupWebSocketEvents();
};

onMounted(() => {
  initializeTerminal();
  xterm.value.open(terminal.value);
  xterm.value.focus();

  setupTerminalEvents();
  initializeWebSocket();
});

useEventListener(window, "resize", () => {
  fitAddon.value.fit();
});

onUnmounted(() => {
  if (isWebSocketOpen()) ws.value.close();
});

defineExpose({ xterm, ws });
</script>

<style scoped lang="scss">
.terminal {
  position: absolute;
  inset: 0;
  margin-right: 0px;
}
</style>
