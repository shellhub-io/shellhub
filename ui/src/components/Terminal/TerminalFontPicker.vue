<template>
  <div class="d-flex flex-column align-center pa-4">
    <v-select
      v-model="currentFontFamily"
      @update:model-value="updateFontSettings"
      :items="availableFonts"
      class="w-100"
      label="Font Family"
    />
    <v-number-input
      v-model="currentFontSize"
      @update:model-value="updateFontSettings"
      control-variant="split"
      variant="outlined"
      density="compact"
      :min="8"
      :max="32"
      hide-details
      inset
      class="w-100"
      label="Font Size"
    />
  </div>
</template>

<script setup lang="ts">
import { storeToRefs } from "pinia";
import useTerminalThemeStore from "@/store/modules/terminal_theme";

const emit = defineEmits<{
  "update:fontSettings": [settings: { fontFamily: string; fontSize: number }];
}>();

const terminalThemeStore = useTerminalThemeStore();
const { availableFonts, currentFontFamily, currentFontSize } = storeToRefs(terminalThemeStore);

const updateFontSettings = async () => {
  await terminalThemeStore.setFontSettings(currentFontFamily.value, currentFontSize.value);
  emit("update:fontSettings", {
    fontFamily: currentFontFamily.value,
    fontSize: currentFontSize.value,
  });
};
</script>
