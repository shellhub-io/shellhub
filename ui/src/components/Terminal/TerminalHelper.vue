<template>
  <BaseDialog v-model="showDialog" @close="close" transition="dialog-bottom-transition" data-test="dialog">
    <v-card class="bg-v-theme-surface">
      <v-card-title class="text-h6 pa-4 bg-primary" data-test="dialog-title">
        Generate a SSH command line
      </v-card-title>

      <v-card-text class="pt-4 pb-0">
        <v-row class="mt-1">
          <v-col>
            <v-text-field
              v-model="username"
              variant="outlined"
              label="Username"
              hint="Enter an existing user on the device"
              persistent-placeholder
              persistent-hint
              density="compact"
              data-test="username-input"
            />
          </v-col>
        </v-row>

        <v-row>
          <v-col>
            <CopyWarning :copied-item="'Command'">
              <template #default="{ copyText }">
                <v-text-field
                  :model-value="commandLine"
                  @click:append-inner="copyText(commandLine)"
                  append-inner-icon="mdi-content-copy"
                  hint="Run this command on your Terminal"
                  class="code"
                  variant="outlined"
                  label="Command Line"
                  readonly
                  density="compact"
                  persistent-placeholder
                  persistent-hint
                  data-test="command-field"
                />
              </template>
            </CopyWarning>
          </v-col>
        </v-row>

        <v-row v-if="showCheckbox">
          <v-col class="d-flex justify-center align-center">
            <v-checkbox
              v-model="dispenseHelper"
              label="Always copy SSHID directly and skip this helper"
              density="compact"
              hide-details
              data-test="dispense-checkbox"
            />
          </v-col>
        </v-row>
      </v-card-text>

      <v-card-actions>
        <v-spacer />
        <v-btn variant="text" data-test="close-btn" @click="showDialog = false">
          Close
        </v-btn>
      </v-card-actions>
    </v-card>
  </BaseDialog>
</template>

<script setup lang="ts">
import { computed, ref, watch } from "vue";
import CopyWarning from "@/components/User/CopyWarning.vue";
import BaseDialog from "../BaseDialog.vue";

interface Props {
  sshid: string;
  userId?: string;
  showCheckbox?: boolean;
}

const props = withDefaults(defineProps<Props>(), {
  userId: "",
  showCheckbox: false,
});

const showDialog = defineModel({ default: false });
const username = ref("");
const dispenseHelper = ref(false);
const LS_KEY = "dispenseTerminalHelper";

const getDispensedUsers = (): string[] => {
  try {
    return JSON.parse(localStorage.getItem(LS_KEY) || "[]");
  } catch {
    return [];
  }
};

const setDispensedUsers = (users: string[]) => {
  localStorage.setItem(LS_KEY, JSON.stringify(users));
};

const commandLine = computed(() => {
  const trimmedUsername = username.value.trim();
  return trimmedUsername ? `ssh ${trimmedUsername}@${props.sshid}` : "";
});

watch(dispenseHelper, (isDispensed) => {
  const users = getDispensedUsers();
  const userIndex = users.indexOf(props.userId);

  if (isDispensed && userIndex === -1) {
    users.push(props.userId);
  } else if (!isDispensed && userIndex !== -1) {
    users.splice(userIndex, 1);
  }

  setDispensedUsers(users);
});

watch(showDialog, (isOpen) => {
  if (isOpen) {
    const users = getDispensedUsers();
    dispenseHelper.value = users.includes(props.userId);
  }
});

const close = () => {
  username.value = "";
  showDialog.value = false;
};

defineExpose({ showDialog });
</script>

<style scoped lang="scss">
.code ::v-deep(.v-field__input) {
  font-family: monospace;
  font-size: 85%;
}
</style>
