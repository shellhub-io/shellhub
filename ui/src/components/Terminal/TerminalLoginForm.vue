<template>
  <FormDialog
    v-model="showDialog"
    @close="handleClose"
    @confirm="submitForm"
    @cancel="handleClose"
    title="Terminal Login"
    icon="mdi-login"
    confirm-text="Connect"
    :confirm-loading="isConnecting"
    :confirm-disabled="!isFormValid"
    cancel-text="Cancel"
    confirm-data-test="submit-btn"
    cancel-data-test="cancel-btn"
  >
    <v-card-text class="d-flex flex-column ga-4 pa-6" data-test="terminal-login-form">
      <v-text-field
        v-model="username"
        :error-messages="usernameError"
        label="Username"
        autofocus
        hint="Enter an existing user on the device"
        persistent-hint
        persistent-placeholder
        :validate-on-blur="true"
        data-test="username-field"
      />

      <v-select
        v-model="authenticationMethod"
        @update:model-value="togglePassphraseField"
        :items="[TerminalAuthMethods.Password, TerminalAuthMethods.PrivateKey]"
        label="Authentication method"
        data-test="auth-method-select"
        hide-details
        class="mb-2"
      />

      <v-select
        v-model="selectedPrivateKeyName"
        @update:model-value="togglePassphraseField"
        v-if="authenticationMethod === TerminalAuthMethods.PrivateKey"
        :items="privateKeysNames"
        item-text="name"
        item-value="data"
        label="Private Key"
        hint="Select a private key file for authentication"
        persistent-hint
        data-test="private-keys-select"
      />

      <v-text-field
        color="primary"
        :append-inner-icon="showPassword ? 'mdi-eye' : 'mdi-eye-off'"
        v-model="password"
        v-else
        :error-messages="passwordError"
        label="Password"
        required
        hint="Enter a valid password for the user on the device"
        persistent-hint
        persistent-placeholder
        data-test="password-field"
        :type="showPassword ? 'text' : 'password'"
        @click:append-inner="showPassword = !showPassword"
        @keydown.enter.prevent="submitForm"
      />

      <v-text-field
        color="primary"
        v-if="showPassphraseField"
        :append-inner-icon="showPassword ? 'mdi-eye' : 'mdi-eye-off'"
        v-model="passphrase"
        :error-messages="passphraseError"
        label="Passphrase"
        required
        hint="Enter the key's passphrase"
        persistent-hint
        persistent-placeholder
        data-test="passphrase-field"
        :type="showPassword ? 'text' : 'password'"
        @click:append-inner="showPassword = !showPassword"
        @keydown.enter.prevent="submitForm"
      />
    </v-card-text>
  </FormDialog>
</template>

<script setup lang="ts">
import { ref, computed } from "vue";
import * as yup from "yup";
import { useField } from "vee-validate";
import { LoginFormData, TerminalAuthMethods } from "@/interfaces/ITerminal";
import { IPrivateKey } from "@/interfaces/IPrivateKey";
import usePrivateKeysStore from "@/store/modules/private_keys";
import FormDialog from "../FormDialog.vue";

const emit = defineEmits<{
  submit: [formData: LoginFormData];
  close: [];
}>();

const showDialog = defineModel<boolean>({ required: true });
const isConnecting = defineModel<boolean>("loading", { default: false });
const { privateKeys } = usePrivateKeysStore();
const authenticationMethod = ref(TerminalAuthMethods.Password);
const showPassword = ref(false);
const selectedPrivateKeyName = ref(privateKeys[0]?.name || "");
const privateKeysNames = privateKeys.map((item: IPrivateKey) => item.name);
const showPassphraseField = ref(false);

const {
  value: username,
  errorMessage: usernameError,
} = useField<string>("username", yup.string().required(), {
  initialValue: "",
});

const {
  value: password,
  errorMessage: passwordError,
} = useField<string>("password", yup.string().required(), {
  initialValue: "",
});

const {
  value: passphrase,
  errorMessage: passphraseError,
  resetField: resetPassphraseField,
} = useField<string>("passphrase", yup.string().required(), {
  initialValue: "",
});

const getSelectedPrivateKey = () => privateKeys.find((item: IPrivateKey) => item.name === selectedPrivateKeyName.value);

const isFormValid = computed(() => {
  if (usernameError.value) return false;

  if (authenticationMethod.value === TerminalAuthMethods.Password) {
    return !passwordError.value && !!username.value && !!password.value;
  }

  if (authenticationMethod.value === TerminalAuthMethods.PrivateKey) {
    if (!selectedPrivateKeyName.value) return false;
    if (showPassphraseField.value && !passphrase.value) return false;
    return !passphraseError.value && !!username.value;
  }

  return false;
});

const togglePassphraseField = () => {
  if (authenticationMethod.value === TerminalAuthMethods.PrivateKey) {
    const hasPassphrase = getSelectedPrivateKey()?.hasPassphrase || false;
    showPassphraseField.value = hasPassphrase;
  } else showPassphraseField.value = false;

  showPassword.value = false;
  resetPassphraseField();
};

const handleClose = () => {
  showDialog.value = false;
  emit("close");
};

const submitForm = () => {
  if (!isFormValid.value) return;

  const privateKey = authenticationMethod.value === TerminalAuthMethods.PrivateKey ? getSelectedPrivateKey()?.data : undefined;

  const formData: LoginFormData = {
    username: username.value,
    password: password.value,
    authenticationMethod: authenticationMethod.value,
    privateKey,
    passphrase: showPassphraseField.value ? passphrase.value : undefined,
  };

  emit("submit", formData);
};

defineExpose({
  authenticationMethod,
  togglePassphraseField,
  isFormValid,
  submitForm,
  username,
  password,
  passphrase,
  selectedPrivateKeyName,
});
</script>
