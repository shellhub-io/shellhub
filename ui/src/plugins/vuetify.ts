// Styles
import "@mdi/font/css/materialdesignicons.css";
import "@fortawesome/fontawesome-free/css/all.css";
import "vuetify/styles";
import { VFileUpload, VFileUploadItem } from "vuetify/labs/VFileUpload";
import "../styles/_variables.scss";
import { aliases as faAliases, fa } from "vuetify/iconsets/fa";

// Vuetify
import { createVuetify } from "vuetify";

const light = {
  dark: false,
  colors: {
    primary: "#667acc",
    secondary: "#FFFFFF",
    background: "#F5F5F5",
    tabs: "#F5F5F5",
    foreground: "#F5F5F5",
    paymentForm: "#F5F5F5",
    terminal: "#0f1526",
    "v-theme-background": "#FFFFFF",
    "v-theme-surface": "#FFFFFF",
    "v-theme-card": "#FFFFFF",
  },
};

const dark = {
  dark: true,
  colors: {
    primary: "#667acc",
    secondary: "#1E2127",
    background: "#18191B",
    tabs: "#1E1E1E",
    foreground: "#1E1E1E",
    paymentForm: "#E0E0E0",
    terminal: "#0f1526",
    "v-theme-background": "#1E2127",
    "v-theme-surface": "#1E2127",
    "v-theme-card": "#22252B",
  },
};

export default createVuetify({
  components: {
    VFileUpload,
    VFileUploadItem,
  },
  theme: {
    defaultTheme: "dark",
    themes: {
      dark,
      light,
    },
  },
  defaults: {
    global: {

    },
    VTextField: {
      variant: "outlined",
      density: "comfortable",
    },
    VSelect: {
      variant: "outlined",
      density: "comfortable",
    },
    VTextarea: {
      variant: "outlined",
    },
  },
  icons: {
    faAliases,
    sets: {
      fa,
    },
  },
});
