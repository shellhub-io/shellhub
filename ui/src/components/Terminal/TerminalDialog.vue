<template>
  <TerminalLoginForm
    v-if="showLoginForm"
    v-model="showLoginDialog"
    v-model:loading="isConnecting"
    @submit="handleSubmit"
    @close="close"
  />

  <Terminal
    v-else
    v-model="showTerminalDialog"
    :key="terminalKey"
    :token="token"
    :private-key="privateKey ?? null"
    :passphrase="passphrase"
    :device-name
    @close="close"
  />
</template>

<script setup lang="ts">
import { ref, watch, computed } from "vue";
import axios from "axios";
import { useRoute, onBeforeRouteLeave } from "vue-router";
import {
  IConnectToTerminal,
  LoginFormData,
  TerminalAuthMethods,
} from "@/interfaces/ITerminal";

// Components used in this dialog
import TerminalLoginForm from "./TerminalLoginForm.vue";
import Terminal from "./Terminal.vue";

// Utility to create key fingerprint for private key auth
import { convertToFingerprint } from "@/utils/sshKeys";

const { deviceUid, deviceName } = defineProps<{
  deviceUid: string;
  deviceName: string;
}>();

const route = useRoute(); // current route
const showLoginForm = ref(true); // controls whether login or terminal is shown
const terminalKey = ref(0);
const showDialog = defineModel<boolean>({ required: true }); // controls visibility of dialog
const isConnecting = ref(false);

const showLoginDialog = computed({
  get: () => showDialog.value && showLoginForm.value,
  set: (value) => { showDialog.value = value; },
});

const showTerminalDialog = computed({
  get: () => showDialog.value && !showLoginForm.value,
  set: (value) => { showDialog.value = value; },
});

// Token and private key values for terminal connection
const token = ref("");
const privateKey = ref<LoginFormData["privateKey"]>("");
const passphrase = ref(); // Passphrase for private key if needed

// Connect to terminal via password or key
const connect = async (params: IConnectToTerminal) => {
  isConnecting.value = true;
  try {
    const response = await axios.post("/ws/ssh", {
      device: deviceUid,
      ...params,
    });

    token.value = response.data.token;
    showLoginForm.value = false;
  } finally {
    isConnecting.value = false;
  }
};

// Handles private key-based connection
const connectWithPrivateKey = async (params: IConnectToTerminal) => {
  const { username, privateKey, passphrase } = params;
  const fingerprint = convertToFingerprint(privateKey as string, passphrase);
  await connect({ username, fingerprint });
};

// Triggered when the user submits login credentials
const handleSubmit = async (params: LoginFormData) => {
  if (params.authenticationMethod === TerminalAuthMethods.Password) {
    await connect(params);
    return;
  }

  await connectWithPrivateKey(params);
  privateKey.value = params.privateKey;
  passphrase.value = params.passphrase || undefined;
  showLoginForm.value = false;
};

// Reset state and close the dialog
const close = () => {
  showDialog.value = false;
  showLoginForm.value = true;
  token.value = "";
  privateKey.value = "";
  terminalKey.value++; // trigger remount
};

// Close terminal and log out when navigating away
onBeforeRouteLeave(() => {
  if (showDialog.value) {
    close();
    return false; // Prevent navigation if dialog is open
  }

  return true;
});

// Auto-open terminal when navigating to specific device route
watch(
  () => route.path,
  (path) => {
    if (path === `/devices/${deviceUid}/terminal`) showDialog.value = true;
  },
  { immediate: true },
);

// Expose for test or parent interaction
defineExpose({
  token,
  handleSubmit,
  showDialog,
  showLoginForm,
  close,
});
</script>
