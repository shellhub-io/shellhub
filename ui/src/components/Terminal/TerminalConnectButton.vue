<template>
  <v-btn
    :disabled="!online"
    :color="online ? 'success' : 'normal'"
    variant="outlined"
    density="comfortable"
    data-test="connect-btn"
    @click="showDialog = true"
  >
    {{ online ? "Connect" : "Offline" }}
  </v-btn>

  <TerminalDialog
    v-model="showDialog"
    :deviceUid
  />
</template>

<script setup lang="ts">
import { ref } from "vue";
import TerminalDialog from "./TerminalDialog.vue";

defineOptions({
  inheritAttrs: false,
});

defineProps<{
  online: boolean;
  deviceUid: string;
}>();

const showDialog = ref(false);

defineExpose({ showDialog });
</script>

<style scoped lang="scss">
button {
  width: min(100%, 120px);
}
</style>
