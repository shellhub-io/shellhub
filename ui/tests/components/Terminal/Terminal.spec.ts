import { mount, VueWrapper } from "@vue/test-utils";
import { createVuetify } from "vuetify";
import { expect, describe, it, beforeEach, vi } from "vitest";
import Terminal from "@/components/Terminal/Terminal.vue";

class MockWebSocket {
  readyState = WebSocket.OPEN;

  send = vi.fn();

  close = vi.fn();

  onopen: (() => void) | null = null;

  onmessage: ((event: { data: string }) => void) | null = null;

  onclose: (() => void) | null = null;
}
vi.stubGlobal("WebSocket", vi.fn(() => new MockWebSocket()));

vi.mock("xterm", () => ({
  Terminal: vi.fn().mockImplementation(() => ({
    open: vi.fn(),
    focus: vi.fn(),
    write: vi.fn(),
    onData: vi.fn(),
    onResize: vi.fn(),
    loadAddon: vi.fn(),
    cols: 80,
    rows: 24,
  })),
}));

describe("Terminal", async () => {
  let wrapper: VueWrapper<InstanceType<typeof Terminal>>;

  const vuetify = createVuetify();

  beforeEach(async () => {
    wrapper = mount(Terminal, {
      global: {
        plugins: [vuetify],
      },
      props: {
        token: "test-token",
      },
    });
  });

  it("Is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("Renders the terminal container", () => {
    expect(wrapper.find("[data-test='terminal-container']").exists()).toBe(true);
  });

  it("initializes WebSocket with correct URL parameters", async () => {
    const mockWsConstructor = vi.mocked(WebSocket);
    expect(mockWsConstructor).toHaveBeenCalledWith(
      expect.stringContaining("/ws/ssh?token=test-token&cols=80&rows=24"),
    );
  });

  it("closes WebSocket connection on component unmount", async () => {
    const mockWs = wrapper.vm.ws;
    wrapper.unmount();
    expect(mockWs.close).toHaveBeenCalled();
  });

  it("initializes xterm with correct configuration", async () => {
    const { Terminal: MockTerminal } = await import("xterm");

    expect(MockTerminal).toHaveBeenCalledWith({
      cursorBlink: true,
      fontFamily: "monospace",
      theme: {
        background: "#0f1526",
      },
    });
  });

  it("loads the FitAddon to xterm", () => {
    const mockXterm = wrapper.vm.xterm;
    expect(mockXterm.loadAddon).toHaveBeenCalledWith(expect.any(Object));
  });

  it("opens xterm in the terminal container", () => {
    const mockXterm = wrapper.vm.xterm;
    const terminalContainer = wrapper.find("[data-test='terminal-container']").element;
    expect(mockXterm.open).toHaveBeenCalledWith(terminalContainer);
  });

  it("focuses on the terminal after initialization", () => {
    const mockXterm = wrapper.vm.xterm;
    expect(mockXterm.focus).toHaveBeenCalled();
  });

  it("sends user input to WebSocket", async () => {
    const mockXterm = wrapper.vm.xterm;
    const mockWs = wrapper.vm.ws as unknown as MockWebSocket;
    const onDataHandler = vi.mocked(mockXterm.onData).mock.calls[0][0];

    onDataHandler("test input");

    expect(mockWs.send).toHaveBeenCalledWith(
      expect.stringContaining('"kind":1'),
    );
    expect(mockWs.send).toHaveBeenCalledWith(
      expect.stringContaining('"data":[116,101,115,116,32,105,110,112,117,116]'),
    );
  });

  it("writes WebSocket data to the terminal", async () => {
    const mockXterm = wrapper.vm.xterm;
    const mockWs = wrapper.vm.ws as unknown as MockWebSocket;

    if (mockWs.onmessage) {
      mockWs.onmessage({ data: "terminal output" });
    }

    expect(mockXterm.write).toHaveBeenCalledWith("terminal output");
  });

  it("handles WebSocket close event", async () => {
    const mockXterm = wrapper.vm.xterm;
    const mockWs = wrapper.vm.ws as unknown as MockWebSocket;

    if (mockWs.onclose) {
      mockWs.onclose();
    }

    expect(mockXterm.write).toHaveBeenCalledWith("\r\nConnection ended");
  });
});
