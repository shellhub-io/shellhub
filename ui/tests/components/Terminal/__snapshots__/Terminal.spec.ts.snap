// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Terminal.vue > renders the terminal container 1`] = `
"<body>
  <div class="v-overlay-container">
    <div class="v-overlay v-overlay--active v-theme--light v-locale--is-ltr v-dialog v-dialog--fullscreen bg-terminal h-100" style="z-index: 2400;" aria-modal="true" role="dialog" data-v-51513c71="">
      <transition-stub name="fade-transition" appear="true" persisted="false" css="true">
        <div class="v-overlay__scrim"></div>
      </transition-stub>
      <transition-stub name="dialog-transition" appear="true" persisted="true" css="true">
        <div class="v-overlay__content" tabindex="-1">
          <div data-v-51513c71="" class="v-card v-theme--light v-card--density-default v-card--variant-elevated bg-v-theme-surface border">
            <!---->
            <div class="v-card__loader">
              <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                <!---->
                <div class="v-progress-linear__background"></div>
                <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                  <div class="v-progress-linear__indeterminate">
                    <div class="v-progress-linear__indeterminate long"></div>
                    <div class="v-progress-linear__indeterminate short"></div>
                  </div>
                </transition-stub>
                <!---->
              </div>
            </div>
            <!---->
            <!---->
            <!-- Content -->
            <!-- Titlebar -->
            <header class="v-toolbar v-toolbar--density-default bg-primary v-theme--light v-locale--is-ltr bg-v-theme-surface border-b px-4 py-2">
              <!---->
              <div class="v-toolbar__content" style="height: 64px;">
                <!---->
                <!---->
                <div class="v-avatar v-theme--light text-primary v-avatar--density-default v-avatar--rounded v-avatar--variant-tonal border border-primary border-opacity-100 mr-3" style="width: 48px; height: 48px;"><i class="mdi-console mdi v-icon notranslate v-theme--light" style="font-size: 24px; height: 24px; width: 24px;" aria-hidden="true"></i>
                  <!----><span class="v-avatar__underlay"></span>
                </div>
                <div>
                  <div class="v-toolbar-title text-h6">
                    <div class="v-toolbar-title__placeholder">
                      <!---->Terminal
                    </div>
                  </div>
                  <div class="text-caption text-medium-emphasis">Connected to test-device</div>
                </div>
                <div class="v-spacer"></div><button type="button" class="v-btn v-btn--icon v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-text" data-test="close-btn-toolbar"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
                  <!----><span class="v-btn__content" data-no-activator=""><i class="mdi-close mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></span>
                  <!---->
                  <!---->
                </button>
                <!---->
              </div>
              <transition-stub name="expand-transition" appear="false" persisted="false" css="true">
                <!---->
              </transition-stub>
            </header><!-- Content -->
            <div data-v-8b2eae94="" class="ma-0 pa-0 w-100 fill-height position-relative">
              <div data-v-8b2eae94="" class="terminal" data-test="terminal-container"></div>
            </div><!-- Footer -->
            <!--v-if-->
            <!---->
            <!----><span class="v-card__underlay"></span>
          </div>
        </div>
      </transition-stub>
    </div>
  </div>
</body>"
`;
